code/modules/mechanics/MechanicMadness.dm

```diff
@@ -1820,6 +1820,195 @@
 		icon_state = "[under_floor ? "u":""]comp_relay"
 		return
 
+#define FIFO_BUFFER 1
+#define FILO_BUFFER 2
+#define RANDOM_BUFFER 3
+#define RING_BUFFER 4 //When I say ring buffer what I really mean is..
+		//...FIFO queue with clobbering like an audio buffer
+
+/obj/item/mechanics/buffercomp
+	name = "Buffer Component"
+	desc = ""
+	icon_state = "comp_buffer"
+	cooldown_time = 0.4 SECONDS
+	var/list/buffer = list()
+	var/buffer_size = 30
+	//I wanted to set this to the same limit as the selection component
+	//But the selection compoment doesn't actually have a limit in the additem() proc
+	//It kind of worries me
+	var/buffer_max_size = 512
+	var/ring_reader = 1
+	var/ring_writer = 1
+
+	var/buffer_model = RING_BUFFER
+	var/buffer_string = "ring"
+	var/buffer_desc = "This mode outputs from oldest to newest, overwriting the oldest signal when full."
+	var/changesig = 0
+
+	get_desc()
+		. += "<br>[SPAN_NOTICE("Delay is [cooldown_time] (in 10ths of a second). <br> Buffer size is [buffer_size].\
+		 <br> Buffer mode is [buffer_string].<br>[buffer_desc]")]"
+
+	New()
+		..()
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_INPUT,"buffer", PROC_REF(buffer))
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_INPUT,"set buffer mode", PROC_REF(compSetModel))
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_CONFIG,"Set Delay",PROC_REF(setDelay))
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_CONFIG,"Set Buffer Mode",PROC_REF(userSetModel))
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_CONFIG,"Set Buffer Size",PROC_REF(setBufferSize))
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ALLOW_MANUAL_SIGNAL)
+		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_CONFIG,"Toggle Signal Changing",PROC_REF(toggleDefault))
+
+	proc/setModel(var/model)
+		//Uppercase everything so signal input can be case insensitive
+		model = uppertext(model)
+		if(model == "FIFO")
+			buffer_desc = "This mode outputs from oldest to newest, dropping signals when full."
+			buffer_model = FIFO_BUFFER
+		else if(model == "FILO")
+			buffer_desc = "This mode outputs from newest to oldest, dropping signals when full."
+			buffer_model = FILO_BUFFER
+		else if(model == "RING")
+			model = "ring"
+			buffer_desc = "This mode outputs from oldest to newest, overwriting the oldest signal when full."
+			buffer_model = RING_BUFFER
+		else if(model == "RANDOM")
+			model = "random"
+			buffer_desc = "This mode outputs signals randomly, randomly overwriting previous signals when full."
+			buffer_model = RANDOM_BUFFER
+		else
+			return FALSE
+
+		buffer_string = model
+		tooltip_rebuild = 1
+		qdel(buffer)
+		buffer = list()
```

Comment:
Much faster way to empty a list than destruction and recreation. We should also probably reset our reader/write indexes
```suggestion
		buffer.len = 0
		ring_reader = 1
		ring_writer = 1
```